### PR TITLE
Adds a test for checking negatives and  NaNs in the temperature field in qv_sat (C++)

### PR DESCRIPTION
**Pull request by singhbalwinder**
This test already exists on the Fortran side, this PR adds this test on the C++ side. 
This PR also completes `bfb_tanh` implementation in the code and  also add a test for `bfb_tanh` function.

### CODE REVIEW

**Comment by singhbalwinder** [comment]
@ambrad and @bartgol : Would you please check if I implemented `bfb_tanh` correctly? 
@PeterCaldwell : I have not pushed the changes for temperature checks yet. I will let you know when I have those working.

**Comment by ambrad** [comment]
Why is there so much redundancy in the tanh decls and impls? Not related to this PR per se, but I thought all of that was supposed to be in files exposing a common set of utils.

For this PR, I don't see the NaN/inf checks.

**Comment by singhbalwinder** [comment]
Thanks @ambrad for your quick review. I basically followed bfb_exp to ensure that I cover all the places. 

I haven't pushed inf/nan changes yet.

**Comment by singhbalwinder** [comment]
I looked over the code again, do you think the definitions in p3 and shoc (on the c++ side) are redundant as we have the same definitions in physcis/common?

**Comment by ambrad** [comment]
Right, I think all redundancy should be removed from the P3 and SHOC files and the common util file included where needed.

**Comment by bartgol** [comment]
I agree, those utils should probably not sit in a specific physics parametrization folder. To be fair, I'm not even sure they belong in `physics/common`; I would put them somewhere in `share`, or even in `scream/ekat/src/ekat/util`. But I guess putting all this stuff in `physics/common` only would already be better.

**Comment by singhbalwinder** [comment]
Since `micro_p3` is using all these functions via `use physics_common_iso_f`, do you think simply removing all these function definitions and impls from C++ P3 and SHOC folders should be sufficient? This way, all these functions will stay in physics/common only and that is what being used in the fortran codes.

**Comment by ambrad** [comment]
Yes.

**Comment by singhbalwinder** [comment]
Ok, I will remove these redundancies from P3 and Shoc.

**Comment by singhbalwinder** [comment]
@ambrad and @bartgol : I have removed the redundant codes and everything passed on my end. I will let AT do one round of testing just to be sure everything is working.

**Comment by singhbalwinder** [comment]
@jgfouca , @ambrad and @bartgol : I think this PR is ready but would you please re-review the changes I introduced?

File: components/scream/src/physics/p3/p3_ice_cldliq_wet_growth_impl.hpp [hunk 1]
```diff
@@ -43,7 +43,7 @@ ::ice_cldliq_wet_growth(
   Spack dum1{0.};
 
   if (any_if.any()) {
-    qsat0 = physics::qv_sat( zerodeg,pres, false );
+    qsat0 = physics::qv_sat( zerodeg,pres, false, range_mask );
```
Comment (ambrad) [inline components/scream/src/physics/p3/p3_ice_cldliq_wet_growth_impl.hpp:46]: Here, T doesn't have to be checked, right?

File: components/scream/src/physics/p3/p3_main_impl.hpp [hunk 2]
```diff
@@ -152,8 +152,8 @@ ::p3_main_part1(
 
     rho(k)     = pdel(k)/dzq(k) / g;
     inv_rho(k) = 1 / rho(k);
-    qvs(k)     = physics::qv_sat(t(k), pres(k), false);
-    qvi(k)     = physics::qv_sat(t(k), pres(k), true);
+    qvs(k)     = physics::qv_sat(t(k), pres(k), false, range_mask);
```
Comment (ambrad) [inline components/scream/src/physics/p3/p3_main_impl.hpp:155]: Here, t will end up being checked twice.

File: components/scream/src/physics/p3/p3_ice_melting_impl.hpp [hunk 3]
```diff
@@ -36,7 +36,7 @@ ::ice_melting(
 
   if (has_melt_qi.any()) {
     //    Note that qsat0 should be with respect to liquid. Confirmed F90 code did this.
-    const auto qsat0 = physics::qv_sat(Spack(Tmelt), pres, false); //last false means NOT saturation w/ respect to ice.
+    const auto qsat0 = physics::qv_sat(Spack(Tmelt), pres, false, range_mask); //"false" here means NOT saturation w/ respect to ice.
```
Comment (ambrad) [inline components/scream/src/physics/p3/p3_ice_melting_impl.hpp:39]: Here again, isn't Tmelt a const? So doesn't need to be checked?

**Comment by ambrad** [comment]
I think you should try to avoid pushing range_mask-type details all the way to the lowest level. Check temperature at the first opportunity. I believe several qv_sat calls are with constants that needn't be checked. If we really need this kind of information going to the bottom-most level -- and there's an argument that we _should_, to maximize good assertions -- we ought to encapsulate this kind of information in a struct that gets passed around, having things in it like nk, range_mask, debug level, and anything else we come up with for diagnosing problems.

**Comment by singhbalwinder** [comment]
Thanks @ambrad for your comments. I wanted to keep the fortran and C++ implementations the same so I added the check for temperature within `qv_sat` on the C++ side as well. As you know, I need range_mask to filter out the padded cells (with uninitialized values) in the packs. @PeterCaldwell would know more but I think we need these checks around saturation vapor pressure calculations as model was crashing during these calculations. Another way is to check the temperature just before the `qv_sat` calls but then we always have to remember that.

I agree with you that it is an overkill to check the temperature in consecutive `qv_sat` calls and also when `qv_sat` is called for a known constant temperature such as `Tmelt`.  I also agree that `range_mask` is more of an infrastructure variable which should not be passed to codes like `qv_sat`. If we need more variables like `range_mask`, we can consider creating a struct and pass that instead of multiple variables.

File: components/scream/src/physics/share/physics_saturation_impl.hpp [hunk 4]
```diff
@@ -11,27 +11,50 @@ namespace physics {
...
+
+  //"range_mask" masks out packs which are padded with uninitialized values    
+  
+  /*Future work:
+  Make error message dynamic so that it tells the user the function name (func_name) from where
+  this error is coming from. Currently func_name is not used in this function */
+
+  const auto is_neg_t_atm = (t_atm <= 0) && range_mask; //find out if there are any negative temperatures in the pack
+  if (is_neg_t_atm.any()){
+    scream_kerror_msg("Error! Temperature has <= 0 values.\n"); //exit with an error message
```
Comment (jgfouca) [inline components/scream/src/physics/share/physics_saturation_impl.hpp:28]: If you use scream_require_msg, you could combine the mask check and error on the same line.

File: components/scream/src/physics/share/physics_saturation_impl.hpp [hunk 5]
```diff
@@ -11,27 +11,50 @@ namespace physics {
...
+  }
+
+  const auto is_nan_t_atm = isnan(t_atm) && range_mask;//find out if there are any NaN temperatures in the pack
+  if (is_nan_t_atm.any()){
+    scream_kerror_msg("Error! Temperature has NaN values.\n"); //exit with an error message
```
Comment (jgfouca) [inline components/scream/src/physics/share/physics_saturation_impl.hpp:33]: Same

File: components/scream/src/physics/share/physics_saturation_impl.hpp (hunk 4 above)
Comment (singhbalwinder) [reply components/scream/src/physics/share/physics_saturation_impl.hpp:28]: When I use `scream_require_msg`, I get the following error:
```
error: device code does not support exception handling
```
`scream_kerror_msg` takes just one argument.

Comment (singhbalwinder) [reply components/scream/src/physics/share/physics_saturation_impl.hpp]: @jgfouca suggested me to use scream_krequire_msg (the device version of this function)

**Comment by AaronDonahue** [comment]
@singhbalwinder , what's the status on this PR?  It looks like there may be some changes requested.  Have you had a chance to take a look?

**Comment by singhbalwinder** [comment]
Yes, I made those changes but since I touched some P3 files, I got some conflicts which I am resolving now. This should be ready to go today (I hope).

**Pull request merged**